### ISSUE TITLE
Mario 3D Land and New Super Mario Bros crash after 1 minute on Mali/Pixel 6

### ISSUE THREAD

**Issue by AlexS2412** [GPUCode/citra#28]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Affected Build(s)

citra-20230301-6e87516 and upward

### Description of Issue

The game runs normally for about 1 minute and crashes with no error. build citra-20230226-6c0aaea is more stable but seems to crash randomly.

### Expected Behavior

The game should just run normally, I'm unsure why these crash as i cannot get it to produce a log.

### Reproduction Steps

Load the game and just try to run it for more than 1 minute.

### Log File

N/A

### System Configuration

CPU: Google Tensor
GPU/Driver: Mali-G78 MP20
RAM: 8GB
OS: Android 12

**Comment by GPUCode** [comment]
Hm could you try the latest commit in vulkan-2? I've solved a small race in the scheduler that might be responsible for this

**Comment by AlexS2412** [comment]
> Hm could you try the latest commit in vulkan-2? I've solved a small race in the scheduler that might be responsible for this

its currently still building it seems. if you wanna chat more directly about this I'm on matrix at https://matrix.to/#/@rinna523:matrix.org

**Comment by AlexS2412** [comment]
i should also mention the screen rotation bug that official citra has. after rotating the screen a few times it can randomly crash.

**Comment by AlexS2412** [comment]
the screen seems to flash blue a lot with wierd distortion for about a frame each time, but it doesn't seem to crash now? partial fix?

**Comment by GPUCode** [comment]
Does the latest build work for you?

**Comment by AlexS2412** [comment]
> Does the latest build work for you?

Sadly it has a different issue, where all shading flickers between on and off, it got introduced in the previous build but wasnt there in citra-20230226-6c0aaea. this regression ends up causing the game to run at 20fps as the flickering happens per frame.

**Comment by AlexS2412** [comment]
not sure if theres a quicker way to debug. it does not crash anymore, but this seizure like effect is probably not a good idea to play with either

**Comment by GPUCode** [comment]
Enable debug renderer and send a logfile with the issue

**Comment by AlexS2412** [comment]
> Enable debug renderer and send a logfile with the issue

i have no clue how to do that... do you not get the issue as well?

**Comment by AlexS2412** [comment]
sorry if I'm not too helpful...

**Comment by GPUCode** [comment]
It's not an issue, I'm glad when people come and report issues. Well to enable debug renderer it's very easy just head over to settings, the debug tab and tick the checkbox named "debug renderer". When you do this games will be much slower but the log will contain additional debug information which could be potentially useful

**Comment by AlexS2412** [comment]
Here is my log from the lastest vulkan build. this build only has the blue flashes occasionally but none of the wierd shading turning on and off like the opengl builds. [Citra Log](https://pastebin.com/TYDGWjp9)

**Comment by AlexS2412** [comment]
seems to be caused by asynchronous shaders... how odd? for games like these its a requirement otherwise they stutter too greatly. but even then the implementation doesn't seem to be working as it should because games still seem to have pauses from shaders if it's enabled... how strange!

**Comment by GPUCode** [comment]
Yes asynchronous shaders can cause a lot of flicker especially on weaker devices that take a long time to compile shaders. My goal is to resolve this with the shader recompiler rewrite but for now my recommendation would be to upgrade your device to Android 13 which supports VK_EXT_pipeline_creation_cache_control. This extensions improves the experience of async shaders

**Comment by GPUCode** [comment]
> captain toad also seems to have horrible graphic distortion even with it off. shadows appear black and there is onjects not cleared from the screen like framebuffer issues when he moves.

Shadow mapping isn't properly implemented but will be very shortly

**Comment by AlexS2412** [comment]
> Yes asynchronous shaders can cause a lot of flicker especially on weaker devices that take a long time to compile shaders. My goal is to resolve this with the shader recompiler rewrite but for now my recommendation would be to upgrade your device to Android 13 which supports VK_EXT_pipeline_creation_cache_control. This extension improves the experience of async shaders

I've been keeping it on 12 because there's no real benefits to android 13 imo, it burns an efuse irreparably because google insisted on ARB with the 13 update, and the developer images aren't quite the same as stable 12. frankly its annoying they act like nintendo and samsung now when it comes to upgrades. i couldn't find even with root a way to bypass ARB.

**Comment by AlexS2412** [comment]
> > Hm could you try the latest commit in vulkan-2? I've solved a small race in the scheduler that might be responsible for this
> 
> its currently still building it seems. if you wanna chat more directly about this I'm on matrix at https://matrix.to/#/@rinna523:matrix.org

also if you could add me on here it may be more convenient to talk directly. im available to test at any time really.

**Comment by AlexS2412** [comment]
also please don't add SAF, it would be a literal nightmare in android 12+ to try to move from internal storage root in android/data because everything is basically locked with SAF and i have my entire 3ds sd card on my phone ;w;

**Comment by GPUCode** [comment]
> also please don't add SAF, it would be a literal nightmare in android 12+ to try to move from internal storage root in android/data because everything is basically locked with SAF and i have my entire 3ds sd card on my phone ;w;

This isn't a fork or citra like MMJ or Enchanced, think of it more like an official approved repo pending for merge. So if mainline citra merges SAF this repo will follow suit. If you have an argument against it make an issue in the main repo. Personally I don't like SAF as well from my limited exposure to it, but it's a necessary evil at this point (all other emus implemented it)

**Comment by GPUCode** [comment]
> > > Hm could you try the latest commit in vulkan-2? I've solved a small race in the scheduler that might be responsible for this
> > 
> > 
> > its currently still building it seems. if you wanna chat more directly about this I'm on matrix at https://matrix.to/#/@rinna523:matrix.org
> 
> also if you could add me on here it may be more convenient to talk directly. im available to test at any time really.

Does that require an account of some sort? You can also join citra's discord server, there's a dedicated channel for vulkan testing and support.

**Comment by AlexS2412** [comment]
> > also please don't add SAF, it would be a literal nightmare in android 12+ to try to move from internal storage root in android/data because everything is basically locked with SAF and i have my entire 3ds sd card on my phone ;w;
> 
> This isn't a fork or citra like MMJ or enhanced, think of it more like an official approved repo pending for merge. So if mainline citra merges SAF this repo will follow suit. If you have an argument against it make an issue in the main repo. Personally I don't like SAF as well from my limited exposure to it, but it's a necessary evil at this point (all other emus implemented it)

mmj offers both still in builds. you dont have to follow suit with bad practice just because everyone else is. and yesh you have to sign up for matrix like anything else really. i just know that i basically wont be able to use newer builds you make if they have SAF because it literally makes it impossible to put anything in android/data, meaning i cant use the citra-emu folder on rhe sd card.

**Comment by GPUCode** [comment]
Uh with SAF you can pick where you store your citra-emu folder so it shouldn't break. You can try the latest canary to confirm this. Unfortunately this PR will get SAF, it has been one of the most requested features for citra and we will get booted from the play store if it isn't implemented.

**Comment by AlexS2412** [comment]
it seems to stll work but async shaders cause the blue flashes on 3d land. it no longer crashes thou!

**Comment by AlexS2412** [comment]
it seems donkey kong country returns also crashes sadly

**Comment by AlexS2412** [comment]
@GPUCode so the games work fine, rotation seemed to be the culprit for why some dont start, which is odd. did you ever figure out the blue screen flashes caused by the asynchronous shaders? was only curious since this game seems to be the only one effected. there is also now a vertical line on the touch screen for OoT3D, like on PC citra. i forget what causes that one though.

**Comment by GPUCode** [comment]
The blue flashes are when a new fragment shader is compiled and a draw is skipped, their frequency can vary between devices and some games exhibit them more than others. Also what vertical line? Can you share a screenshot of that?

**Comment by GPUCode** [comment]
Does this still happen with the newer vulkan builds?

**Comment by AlexS2412** [comment]
yup. no change. also long load times because of SAF.
